File: java/extentreports/src/main/java/com/relevantcodes/extentreports/converters/TestConverter.java
package com.relevantcodes.extentreports.converters;

import java.io.File;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import com.relevantcodes.extentreports.ExtentReports;
import com.relevantcodes.extentreports.ExtentTest;
import com.relevantcodes.extentreports.LogSettings;
import com.relevantcodes.extentreports.model.Log;
import com.relevantcodes.extentreports.model.Test;
import com.relevantcodes.extentreports.utils.DateTimeUtil;
import com.relevantcodes.extentreports.utils.FileReaderEx;

public class TestConverter extends LogSettings {
	private static final Logger logger = Logger.getLogger(TestConverter.class.getName());
	private Document doc;
	private File file;
	private ExtentReports extent;
	
	public void createTestList() {
		Elements body = doc.select("body.extent");
		
		if (body == null || body.size() == 0) {
			logger.log(
					Level.SEVERE, 
					"The supplied file " + file.getAbsolutePath() + " is not a valid Extent file. " +
					"Parsing failed, tests from the supplied file will not be listed in the current report."
			);
			
			return;
		}
		
		Elements allTests = doc.select(".test");
		
        ExtentTest extentTest;
		
        // build Test model
        for (Element test : allTests) {
            String name = test.select(".test-name").first().html().trim();
            String description = test.select(".test-desc").first().html().trim();
            
            extentTest = extent.startTest(name, description);
            
            extentTest.getTest().setStartedTime(
            		DateTimeUtil.getDate(
            				test.select(".test-started-time").first().text(), 
            				LogSettings.getLogDateTimeFormat()
            		)
            );
            extentTest.getTest().setEndedTime(
            		DateTimeUtil.getDate(
            				test.select(".test-ended-time").first().text(),
            				LogSettings.getLogDateTimeFormat()
            		)
            );
            
            Elements categories = test.select(".category");
            
            for (Element category : categories) {
            	extentTest.assignCategory(category.text());
            }
            
            Elements authors = test.select(".author");
            
            for (Element author : authors) {
            	extentTest.assignAuthor(author.text());
            }
            
            LogConverter logConverter = new LogConverter(test);
            List<Log> logList = logConverter.getLogList();
            
            if (logList != null && logList.size() > 0) {
            	extentTest.getTest().setLog(logList);
            }
            
            ChildTestConverter nodeConverter = new ChildTestConverter(test);
            List<Test> nodeList = nodeConverter.getNodeList();
            
            if (nodeList != null && nodeList.size() > 0) {
            	extentTest.getTest().hasChildNodes = true;
            	extentTest.getTest().setNodeList(nodeList);
            }
            
            extent.endTest(extentTest);
        }
    }
	
	public TestConverter(ExtentReports extent, File file) {
		this.file = file;
		this.extent = extent;
		
		String source = FileReaderEx.readAllText(file);
		doc = Jsoup.parse(source);
	}
}

File: dotnet/RelevantCodes.ExtentReports/RelevantCodes.ExtentReports/Converters/TestConverter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using HtmlAgilityPack;

using RelevantCodes.ExtentReports.Model;

namespace RelevantCodes.ExtentReports.Converters
{
    class TestConverter : IConvertible<ExtentTest>
    {
        private ExtentReports _extent;
        private string _filePath;

        public TestConverter(ExtentReports Extent, string FilePath)
        {
            _extent = Extent;
            _filePath = FilePath;
        }

        public List<ExtentTest> Convert()
        {
            var document = new HtmlAgilityPack.HtmlDocument();
            document.Load(_filePath);

            var documentNode = document.DocumentNode;

            var testCollectionNode = documentNode.Descendants("ul").Where(x => x.Id == "test-collection");

            if (testCollectionNode != null && testCollectionNode.Any() && testCollectionNode.First().ChildNodes.Any())
            {
                var tests = testCollectionNode.First().ChildNodes.Where(x => x.Name == "li");

                var list = new List<ExtentTest>();

                tests.ToList().ForEach(htmlTestNode => 
                {
                    var test = HTMLNodeToExtentTest(htmlTestNode);

                    _extent.EndTest(test);
                    list.Add(test);
                });

                return list;
            }

            return null;
        }

        private ExtentTest HTMLNodeToExtentTest(HtmlNode HtmlTestNode)
        {
            var testName = HtmlTestNode.Descendants("span")
                    .Where(x => x.Attributes["class"].Value.Contains("test-name"))
                    .First().InnerHtml;

            var description = HtmlTestNode.Descendants("div")
                .Where(x => x.Attributes["class"].Value.Contains("test-desc"))
                .First().InnerHtml;

            var extentTest = _extent.StartTest(testName, description);
            extentTest.GetTest().ID = new Guid(HtmlTestNode.Attributes["extentid"].Value);

            var status = HtmlTestNode.Descendants("span")
                .Where(x => x.Attributes["class"].Value.Contains("test-status"))
                .First().InnerText;
            
            var startTime = HtmlTestNode.Descendants("span")
                .Where(x => x.Attributes["class"].Value.Contains("test-started-time"))
                .First().InnerText;
            extentTest.GetTest().StartTime = System.Convert.ToDateTime(startTime);

            var endTime = HtmlTestNode.Descendants("span")
                .Where(x => x.Attributes["class"].Value.Contains("test-ended-time"))
                .First().InnerText;
            extentTest.GetTest().EndTime = System.Convert.ToDateTime(endTime);

            var categories = HtmlTestNode.Descendants("span")
                .Where(x => x.Attributes["class"].Value.Contains("category"));
            var authors = HtmlTestNode.Descendants("span")
                .Where(x => x.Attributes["class"].Value.Contains("author"));
            AssignTestAttributes(categories, authors, extentTest);

            var testStepsNode = HtmlTestNode.Descendants("div").Where(x => x.Attributes["class"].Value.Contains("test-steps"));
            ParseAndAddLogs(testStepsNode, extentTest);

            var nodeListNode = HtmlTestNode.Descendants("ul").Where(x => x.Attributes["class"].Value.Contains("node-list"));
            ParseAndAddNodes(nodeListNode, extentTest);

            return extentTest;
        }

        private void AssignTestAttributes(IEnumerable<HtmlNode> CategoriesNode, IEnumerable<HtmlNode> AuthorsNode, ExtentTest ExtentTest)
        {
            CategoriesNode.ToList().ForEach(cat =>
            {
                if (!string.IsNullOrEmpty(cat.InnerText.Trim()))
                {
                    ExtentTest.AssignCategory(cat.InnerText);
                }
            });

            AuthorsNode.ToList().ForEach(author =>
            {
                if (!string.IsNullOrEmpty(author.InnerText.Trim()))
                {
                    ExtentTest.AssignAuthor(author.InnerText);
                }
            });
        }

        private void ParseAndAddLogs(IEnumerable<HtmlNode> TestStepsNode, ExtentTest ExtentTest)
        {
            if (TestStepsNode != null & TestStepsNode.Any())
            {
                if (TestStepsNode.First().ChildNodes.Where(x => x.Name.Equals("table")).Count() == 0)
                {
                    return;
                }

                var logs = new LogConverter(TestStepsNode.First()).Convert();

                if (logs != null)
                {
                    ExtentTest.GetTest().LogList = logs;
                }
            }
        }

        private void ParseAndAddNodes(IEnumerable<HtmlNode> NodeListNode, ExtentTest ExtentTest)
        {
            if (NodeListNode != null & NodeListNode.Any())
            {
                ExtentTest.GetTest().ContainsChildNodes = true;

                var nodes = new NodeConverter(NodeListNode.First(), ExtentTest).Convert();

                if (nodes != null)
                {
                    ExtentTest.GetTest().NodeList.AddRange(nodes);
                }
            }
        }
    }
}
